## Q1

In [1]:
import pandas as pd
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
df = pd.read_csv(url)
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Q2

After Trying to acomplishing task 1 in prelecture HW, Chatgpt gave me a code to run as anwser to my question since it couldn't open the dataset himeself

here is the code:

In [3]:
import pandas as pd

# Load the Titanic dataset from your local file
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Get the columns of the dataset and the size (rows, columns)
columns = titanic_data.columns
size = titanic_data.shape

# Display the columns and the size
print(f"Columns: {columns}")
print(f"Size: {size}")

Columns: Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
Size: (891, 12)


# Q 2.2

Well to be fair GPT gave me a really well rounded anwser but I think i deserve some creddit for good prompts =)

in nthe context of my dataset: 

- Observations = rows = individual passengers.
- Variables = columns = characteristics or data points about each passenger.

I discussed the output with ChatGPT, since I didn't exactly know what "Size: (891, 12)" stood for, the data set -even though with some missing values- is a dataset of 891 Titanic passengers and 12 characteristics of each individual represented in each column

## Q3

After someback and forth between me and GPT we landed on a decent qway to sumrize the ccolumns

1. For numerical columns, I'll provide basic statistics like mean, min, max, and missing values.
2. For categorical columns (like Sex, Embarked), I'll show the distribution of unique values and missing data.

note that in the list above I refers to ChatGPT

One neat problem was the Survived column which was technicly numeric but it used 0s and 1s as boolians

simllarly to one before, Pclass is only 1s 2s and 3s, again technicly numeric but it refers to Classes not actual numbers

I asked GPT to adjust for thoose 

Here is the code:

In [6]:
import pandas as pd

# Load the Titanic dataset from your local file
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Basic summary for each column including missing values
def summarize_column(col):
    print(f"\nSummary for column: {col.name}")
    print(f"  Data type: {col.dtype}")
    print(f"  Missing values: {col.isnull().sum()}")

    # Special handling for Survived (binary column)
    if col.name == 'Survived':
        print(f"  Survival distribution:\n{col.value_counts()}")
        survival_rate = col.mean() * 100
        print(f"  Survival rate: {survival_rate:.2f}%")
    
    # Special handling for Pclass (categorical class column)
    elif col.name == 'Pclass':
        print(f"  Passenger Class distribution:\n{col.value_counts()}")
    
    # If numeric, summarize like normal numeric data
    elif pd.api.types.is_numeric_dtype(col):
        print(f"  Mean: {col.mean():.2f}")
        print(f"  Min: {col.min()}")
        print(f"  Max: {col.max()}")
        print(f"  Standard Deviation: {col.std():.2f}")
    else:
        # Otherwise, treat it as categorical and show unique values
        print(f"  Unique values:\n{col.value_counts(dropna=False)}")

# Loop through all columns in the dataset and summarize
for col_name in titanic_data.columns:
    summarize_column(titanic_data[col_name])


Summary for column: PassengerId
  Data type: int64
  Missing values: 0
  Mean: 446.00
  Min: 1
  Max: 891
  Standard Deviation: 257.35

Summary for column: Survived
  Data type: int64
  Missing values: 0
  Survival distribution:
Survived
0    549
1    342
Name: count, dtype: int64
  Survival rate: 38.38%

Summary for column: Pclass
  Data type: int64
  Missing values: 0
  Passenger Class distribution:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Summary for column: Name
  Data type: object
  Missing values: 0
  Unique values:
Name
Braund, Mr. Owen Harris                     1
Boulos, Mr. Hanna                           1
Frolicher-Stehli, Mr. Maxmillian            1
Gilinski, Mr. Eliezer                       1
Murdlin, Mr. Joseph                         1
                                           ..
Kelly, Miss. Anna Katherine "Annie Kate"    1
McCoy, Mr. Bernard                          1
Johnson, Mr. William Cahoone Jr             1
Keane, Miss. Nora A             

## Q4

Once again, Hats off to GPT for giving such complite anwser im basicly just writing it in md, but here it is:

- **Non-numeric variables are excluded from df.describe()**
    - If you have 12 columns but 5 of them are non-numeric, df.describe() will only analyze the remaining 7 numeric columns, leading to fewer columns in the describe() output.
    - df.describe(), by default, only analyzes numeric variables, so it will exclude non-numeric (categorical) columns like Name, Sex, Ticket, Cabin


- **Missing numeric values will not be analyized by df.describe()**
    - df.describe() gives summaries like mean, min, max, and counts only non-missing (valid) entries for each numeric column.
    - in casse of our dataset, the analisis above is only for existing ages since our set has a lot of missing ages, the min max and mean are for only the valid ones
    
- df.shape reports the total number of rows and columns in your dataset.
    
to oversimlify it but in my own words, in case of non-numeric vars, df.describe() will not analyze and in case of missing numeric vars df.describe() will only count the values that are there. To the contrary df.shape will count every row and column regardless of it being numeric or not and missing values or not it will give the overall number of rows and columns

## Q5

#### Method:
- A method is a **function** that is associated with an object, and it can perform operations or actions on the object’s data, typically modifying or providing a new view of the data.

#### Attribute:
- An attribute is a **property** of an object that stores data or state information about the object.

above you see the gpt responce but it think it's best said that an atribiute like shape is something that in oop is associathed with our object at initilazation and the maybe changed but a method is a functtion that we can use on our object.

methods might take arguments and perform an operation but attribut are accessed directly from the object itself.

### Pre-lecture HW, Chatbot summary 

Here is the link to my ChatGPT intraction :
https://chatgpt.com/share/dcb58cec-d4bd-4c51-8b26-59c131f95bad

[Click Here](https://chatgpt.com/share/dcb58cec-d4bd-4c51-8b26-59c131f95bad)

---

#### Understanding the Titanic Dataset Structure

Objective: To gain a foundational understanding of the Titanic dataset's structure and how to interpret its dimensions and variables.

#### Dataset Dimensions:

- Rows (Observations): Each of the 891 rows represents an individual passenger aboard the Titanic.
- Columns (Variables): The 12 columns represent different characteristics (variables) describing each passenger (e.g., PassengerId, Survived, Pclass, Age).

#### Explaining Observations and Variables:

- Observations: Refers to individual entries in the dataset; in this case, individual Titanic passengers.
- Variables: Refers to the features or attributes recorded for each observation, such as survival status, class, and age.

#### Discrepancies in Data Analysis:

- Non-Numeric Variables: df.shape reports all columns, but df.describe() by default only includes numeric variables. This results in fewer columns being analyzed by df.describe().
- Missing Values in Numeric Variables: df.shape counts all rows, but df.describe() counts only non-missing entries for each column. This can lead to discrepancies in the "count" between these two methods, particularly noticeable in columns with missing data like Age.

#### Summary Statistics:

Discussed how to generate and interpret summary statistics for both numeric and categorical data within the dataset, adjusting the approach based on the nature of each variable (e.g., treating Survived as a categorical variable).

#### Code Adjustments:
Adjusted Python code to correctly summarize the dataset, accounting for the specific nature of binary and categorical columns like Survived and Pclass.

#### Conclusion:
The Titanic dataset was dissected to understand its structure and the characteristics of its variables. Practical steps to summarize the dataset were discussed, and Python code was provided and adjusted to handle specific data types accurately.